### PR TITLE
Feature render as table or  json

### PR DESCRIPTION
**Pull request by c0dehard**
#### Mistakes removed, update to match for gsclient 3.0.1:
- special thanks goes to @nvthongswansea 
___
#### new commands:
    gscloud server
    gscloud storage
    gscloud network
#### also new:
    global flag --json -j
example command for pretty json output:
    ./gscloud server -j | jq

### CODE REVIEW

File: cmd/makeConfig.go [hunk 1]
```diff
@@ -59,7 +59,6 @@ func emptyConfig() []byte {
 		Name:   "default",
 		UserID: "",
 		Token:  "",
-		URL:    defaultAPIURL,
```
Comment (bkircher) [inline cmd/makeConfig.go:62]: Why is this gone?

File: cmd/cliconfig.go [hunk 2]
```diff
@@ -61,14 +60,20 @@ func newCliClient(account string) *gsclient {
...
-	}
-	return newClient(clientConf)
+	// clientConf := &clientConfig{
+	// apiURL:     defaultAPIURL,
+	// userUUID:   ac.UserID,
+	// userToken:  ac.Token,
+	// userAgent:  "gscloud",
+	// httpClient: http.DefaultClient,
+	// }
+	// return newClient(clientConf)
```
Comment (bkircher) [inline cmd/cliconfig.go:70]: Do not commit commented out code. Remove it.

Here and other hunks.

File: cmd/cliconfig.go [hunk 3]
```diff
@@ -14,7 +14,6 @@ type accountEntry struct {
 	Name   string `yaml:"name"`
 	UserID string `yaml:"userId"`
 	Token  string `yaml:"token"`
-	URL    string `yaml:"url"`
```
Comment (bkircher) [inline cmd/cliconfig.go:17]: As with the change in the make-config subcommand, what's the reason for this change?

File: cmd/root.go [hunk 4]
```diff
@@ -34,10 +36,12 @@ func init() {
```
Comment (bkircher) [inline cmd/root.go:36]: Following hunks are just unnecessary white space changes. They are just noise. Remove them.

File: cmd/root.go [hunk 5]
```diff
@@ -34,10 +36,12 @@ func init() {
 
 	rootCmd.PersistentFlags().StringVar(&cfgFile, "config", "", fmt.Sprintf("configuration file, default %s", cliConfigPath()))
 	rootCmd.PersistentFlags().StringVar(&account, "account", "", "the account used, 'default' if none given")
+	rootCmd.PersistentFlags().BoolVarP(&jsonFlag, "json", "j", false, "Returns a table as json.")
```
Comment (bkircher) [inline cmd/root.go:39]: Maybe better "Print results as JSON"" or "Print output as JSON". Also note: flags and help texts don't end with a trailing dot.

File: cmd/network.go [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+	"github.com/gridscale/gscloud/render"
+	"github.com/spf13/cobra"
+)
+
+// networkCmd represents the network command
+var networkCmd = &cobra.Command{
+	Use:   "network",
+	Short: "Print network list",
+	Long: `Display a network list as table by default 
+	as json by using the flag --json or -j`,
```
Comment (bkircher) [inline cmd/network.go:17]: No need to doc global flags here. Just write what the subcommand does. E.g., 
Short: "Print networks"
Long: "Print all networks along with their... whatever"

File: cmd/cliconfig.go (hunk 2 above)
Comment (c0dehard) [reply cmd/cliconfig.go:70]: removed

File: cmd/cliconfig.go (hunk 3 above)
Comment (c0dehard) [reply cmd/cliconfig.go:17]: Added again.
Yes because I used another url, had to do some changes to test the build

File: cmd/root.go (hunk 4 above)
Comment (c0dehard) [reply cmd/root.go:36]: removed

File: cmd/root.go (hunk 5 above)
Comment (c0dehard) [reply cmd/root.go:39]: fixed

File: cmd/network.go (hunk 6 above)
Comment (c0dehard) [reply cmd/network.go:17]: changed text

File: cmd/makeConfig.go (hunk 1 above)
Comment (c0dehard) [reply cmd/makeConfig.go]: Forgot to reimplement this, had to remove because i used a different url (to test my build with dev faster)

Comment (c0dehard) [reply cmd/makeConfig.go]: And because there's no longer need of client.go, but as mentioned its reimplemented now

File: cmd/root.go (hunk 5 above)
Comment (bkircher) [reply cmd/root.go]: Ehm, no. I meant the other way around. Sorry my wording wasn't very clear.

Flags and help texts should _not_ end with a trailing dot.

So, no dot.

**Comment by bkircher** [comment]
Okay, pretty cool. This works already
```
$ ./gscloud server ls --json|jq
[
  [
    {
      "server": {
        "object_uuid": "8bbc3917-a0c0-42c6-b3b7-2087cc33cb91",
        "name": "s1",
        "memory": 1,
        "cores": 1,
        "hardware_profile": "default",
        "status": "active",

// …
```
Awesome.


Couple of things to do:

1. There are still commented out lines in the diff. Please remove all of them. To get a meaningful diff you can do something like this:

```
    $ git diff develop HEAD --name-only | grep -v '^vendor/' | xargs git diff develop HEAD --
```

2. POWER column shows true or false in the table. Can you make it on or off instead?

```
$ ./gscloud server ls

SERVER  SPECIFICATIONS  POWER  CURRENTPRICE  
s1      Core/s 1 RAM 1  false  0             
s2      Core/s 1 RAM 2  false  0             
s3      Core/s 1 RAM 2  true   0          
```

3. Split up SPECIFICATIONS column into two MEM and CORE and display only numbers in the rows?

4. There is a blank line at the start in the output. Remove it. Should look like this:
```
$ ./gscloud server ls
SERVER  CORE MEM  POWER
s1      1    1     off
s2      1    2     off 
s3      1    2     on  
```

5. remove CURRENTPRICE for now from table

**Comment by c0dehard** [comment]
The [**nl**](https://www.computerhope.com/jargon/n/newline.htm) issue was caused by github.com/rodaine/table/table.go : line 209 

[Figured out the issue was already known](https://github.com/rodaine/table/issues/12#issue-314332218)

```{.go .numberLines startFrom="206"}
func (t *table) Print() {
	format := strings.Repeat("%s", len(t.header)) + "\n"
	t.calculateWidths()
	fmt.Fprintln(t.Writer)
	t.printHeader(format)
	for _, row := range t.rows {
		t.printRow(format, row)
	}
}
```

```shell
$ ./command table 

SERVER  CORE  MEM  POWER  
s1      1     1    off    
s2      1     2    off    
```

___



##### So i  just removed it

```{.go .numberLines startFrom="206"}
func (t *table) Print() {
	format := strings.Repeat("%s", len(t.header)) + "\n"
	t.calculateWidths()
	t.printHeader(format)
	for _, row := range t.rows {
		t.printRow(format, row)
	}
}
```

```shell
$ ./command table 
SERVER  CORE  MEM  POWER  
s1      1     1    off    
s2      1     2    off    
```

**Comment by bkircher** [comment]
> So i just removed it

Yeah, but this is not a change that can be merged. Anyway lets postpone this requested change for later.

As for the other changes. Please make sure that your commits follow https://chris.beams.io/posts/git-commit/#seven-rules. (Remember, it is your PR. This means you want to integrate your changes into this project. So it is your responsibility to make this PR as cleanly reviewable and understandable as possible.)

Idea: maybe squash this branch and force push again and add a comment with open things that need to be done and change requests that have already be done.
